### ISSUE TITLE
how to trace & tune ?

### ISSUE THREAD

**Issue by tissatussa** [mhib/combusken#7]
in your readme i see info about "combusken tune" and "combusken trace-tune" .. it's not clear to me how this works : when compiling, should "tune" be done to output a better binary ?  Same for trace-tune, it's not clear how this works and how it should be executed .. and what (type of) chess games can games.fen hold for tuning ? Can a special binary be compiled this way, eg. for Sicilian Dragon opening positions ? Can you supply an example game.fen file ?

**Comment by mhib** [comment]
Hi!

Sorry for the late answer.

These are 2 implementations of [Texel's Tuning Method](https://www.chessprogramming.org/Texel%27s_Tuning_Method).

* `combusken tune` directly manipulates weights and uses coordinate descent combined with gradient descent(gradient calculated with symmetric derivative) as an optimizer
* `combusken trace-tune` uses trace to remember which weights were used for a given position and uses Adagrad. 


As for generating binary; the current process is that the tuner only prints out weights at the ends of tuning, then the weights are replaced using script in `tools/replace_values_in_eval_with_results.rb`.  It is possible to generate a training set that makes Combusken prefer Sicilian Dragon to other openings if that's the question. Note that this is not special for Combusken, and it would work with any other chess engine that supports evaluation weights tuning. I have no expertise in generating "prefer-this-opening" training sets.
<hr>
games.fen has this format:

```
rnb1kbnr/pp1pppp1/7p/2q5/5P2/N1P1P3/P2P2PP/R1BQKBNR w KQkq - c9;1/2-1/2
r1q1kb1r/6pp/b1p1pn2/2P1Np2/QP6/4P3/P2N2PP/R1BR2K1 b kq - c9;1/2-1/2
8/8/1p1k4/1P6/8/3p3P/1r4P1/5K2 w - - c9;0-1
r6r/n6p/6p1/1p3pb1/q7/P1B1kP1P/6P1/6K1 b - - c9;0-1
8/6P1/8/1pk1r3/6RP/8/3K4/8 b - - c9;1-0
```
